framework/include/utils/MathUtils.h

```diff
@@ -198,4 +200,20 @@ smootherStep(T x, T2 start, T2 end, bool derivative = false)
   return Utility::pow<3>(x) * (x * (x * 6.0 - 15.0) + 10.0);
 }
 
+template <typename T>
+Real
+sqrt(T x)
+{
+  mooseAssert(x >= 0.0, "Cannot take square root of negative value");
+  return std::sqrt(x);
+}
+
+template <ComputeStage compute_stage>
+ADReal
+sqrt(ADReal x)
+{
+  mooseAssert(x >= 0.0, "Cannot take square root of negative value");
+  return std::sqrt(x);
```

Comment:
I can't get a protected sqrt to work for ADReal. Any suggestions @lindsayad @dschwen ?

Reply:
What do you mean by can't get it to work? What's the error? Anyway I would change this to not be templated at all, because you already have the general `T` template. I would do:
```c++
DualReal
sqrt(const DualReal & x)
{
  mooseAssert()
  return std::sqrt(x);
}
```
where the only reason for this overload is to prevent copy construction of the `sqrt` argument

Reply:
This had to be templated due to the dualnumber nature of DualReal...let me find the error...

Reply:
`DualReal` does not take any template arguments. `ADReal` does require a template argument. I switched from `ADReal` to `DualReal` in the example code I wrote

Reply:
I must be missing an include or something?

<details>
<summary>
error
</summary>

```
Compiling C++ (in opt mode) /Users/topher/projects/moose/framework/src/utils/BicubicInterpolation.C...
In file included from /Users/topher/projects/moose/framework/src/utils/BicubicInterpolation.C:12:
/Users/topher/projects/moose/framework/build/header_symlinks/MathUtils.h:237:1: error: implicit instantiation of undefined template 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned int,
      MetaPhysicL::NWrapper<50> > >'
sqrt(const DualReal & x)
^
/Users/topher/projects/moose/framework/build/header_symlinks/DualReal.h:20:7: note: template is declared here
class DualNumber;
      ^
In file included from /Users/topher/projects/moose/framework/src/utils/BicubicInterpolation.C:12:
/Users/topher/projects/moose/framework/build/header_symlinks/MathUtils.h:240:10: error: no matching function for call to 'sqrt'
  return std::sqrt(x);
         ^~~~~~~~~
/usr/include/math.h:433:15: note: candidate function not viable: cannot convert argument of incomplete type 'const DualReal' (aka 'const DualNumber<double, SemiDynamicSparseNumberArray<double, unsigned int, NWrapper<50> > >') to 'double' for 1st
      argument
extern double sqrt(double);
              ^
/opt/moose/llvm-8.0.0/bin/../include/c++/v1/math.h:1031:46: note: candidate function not viable: cannot convert argument of incomplete type 'const DualReal' (aka 'const DualNumber<double, SemiDynamicSparseNumberArray<double, unsigned int,
      NWrapper<50> > >') to 'float' for 1st argument
inline _LIBCPP_INLINE_VISIBILITY float       sqrt(float __lcpp_x) _NOEXCEPT       {return ::sqrtf(__lcpp_x);}
                                             ^
/opt/moose/llvm-8.0.0/bin/../include/c++/v1/math.h:1032:46: note: candidate function not viable: cannot convert argument of incomplete type 'const DualReal' (aka 'const DualNumber<double, SemiDynamicSparseNumberArray<double, unsigned int,
      NWrapper<50> > >') to 'long double' for 1st argument
inline _LIBCPP_INLINE_VISIBILITY long double sqrt(long double __lcpp_x) _NOEXCEPT {return ::sqrtl(__lcpp_x);}
                                             ^
/opt/moose/llvm-8.0.0/bin/../include/c++/v1/math.h:1038:1: note: candidate template ignored: requirement 'std::is_integral<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned int, MetaPhysicL::NWrapper<50> > >
      >::value' was not satisfied [with _A1 = MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned int, MetaPhysicL::NWrapper<50> > >]
sqrt(_A1 __lcpp_x) _NOEXCEPT {return ::sqrt((double)__lcpp_x);}
^
/opt/moose/llvm-8.0.0/bin/../include/c++/v1/complex:1044:1: note: candidate template ignored: could not match 'complex' against 'DualNumber'
sqrt(const complex<_Tp>& __x)
^
2 errors generated.
```
</details>

Reply:
Since the function I wrote out for you is not templated, you will actually need to just declare it in the header and then define it in `MathUtils.C` in order to avoid violating the One Definition Rule. And then in `MathUtils.C` make sure to have `DualRealOps.h` as an include

Reply:
Ah yes, of course. Well thanks for closing the circle on how to get this to work, but I'll leave this be for now.